File: tests/fixtures/tauri-events-dynamic/frontend/src/events.ts
// The frontend half of the event protocol.
//
// Every string here that names an event is also a string on the Rust side,
// except the two that are internal to one language — those are marked.

import { emit, emitTo, listen, once } from "@tauri-apps/api/event";
import { getCurrentWebviewWindow } from "@tauri-apps/api/webviewWindow";

/// A receiver that is not a Tauri window, and a bus that is not Tauri's.
///
/// These exist to be ignored. `emit` and `listen` are ordinary names, and a
/// pass that matched on the callee alone would take these for Tauri calls.
interface Bus {
  emit(name: string): void;
  once(name: string): void;
}
declare const socket: Bus;
declare const emitter: Bus;

export function notTauri(): void {
  // If either of these were mistaken for a Tauri call, the event they name
  // would gain a producer or a consumer it does not have, and stop being
  // classified as external.
  socket.emit("plugin-status");
  emitter.once("telemetry-ping");
}

/// Emitted here, listened for in Rust.
export async function announceReady(): Promise<void> {
  await emit("frontend-ready", { at: Date.now() });
}

/// `emitTo` takes a window label first. The label is not this pass's business.
export async function pushState(): Promise<void> {
  await emitTo("main", "sync-state", { ready: true });
}

export function watch(): void {
  void listen("download-progress", () => {});
  void listen<{ user: string }>("session-updated", () => {});
  void once("startup-complete", () => {});
}

/// The name held in a module constant.
const EVENT = "open-file";

export function watchOpenFile(): void {
  void listen(EVENT, () => {});
}

/// Emitted and listened for entirely inside the frontend.
export async function refresh(): Promise<void> {
  await emit("fe-internal-refresh", {});
}

export function onRefresh(): void {
  void listen("fe-internal-refresh", () => {});
}

/// Through a window receiver, rather than the module functions.
export async function resized(): Promise<void> {
  const win = getCurrentWebviewWindow();
  await win.emit("window-resized", { w: 1, h: 1 });
}

export function onResized(): void {
  const win = getCurrentWebviewWindow();
  void win.listen("window-resized", () => {});
}

/// A runtime-computed event name on the frontend side.
export function wildcard(eventName: string): void {
  void listen(eventName, () => {});
}

/// Only ever listened for. Something outside this workspace produces it — a
/// plugin, or another page — so the name stays.
export function watchPlugin(): void {
  void listen("plugin-status", () => {});
}

File: tests/fixtures/tauri-events-static/src-tauri/src/events.rs
//! Every Rust-side event shape Tauri 2 offers.
//!
//! Each event name below is also a string on the frontend side, except for the
//! two that are internal to one language — those are marked.

use tauri::{AppHandle, Emitter, Listener};

/// A type with an `emit` and a `listen` of its own.
///
/// Nothing says these names belong to Tauri, and a pass that matched on the
/// method name alone would rewrite this call. It must not.
pub struct LocalBus;

impl LocalBus {
    pub fn emit(&self, _name: &str, _payload: u32) {}
    pub fn listen(&self, _name: &str) {}
}

pub fn not_tauri(bus: &LocalBus) {
    // If either of these were taken for a Tauri call, the event it names would
    // gain a producer or a consumer it does not have, and would stop being
    // classified as reaching outside the workspace.
    bus.emit("plugin-status", 1);
    bus.listen("telemetry-ping");
}

pub fn produce<R: tauri::Runtime>(app: &AppHandle<R>) {
    // Cross-language: the frontend listens for both of these.
    app.emit("download-progress", 1u32).unwrap();
    app.emit_to("main", "session-updated", 2u32).unwrap();
    app.emit_filter("open-file", 3u32, |_target| true).unwrap();

    // Listened for by the frontend only.
    app.emit("startup-complete", 4u32).unwrap();

    // Rust to Rust.
    app.emit("rust-internal-tick", 5u32).unwrap();

    // Nothing in the workspace listens for this, so something outside it may.
    app.emit("telemetry-ping", 6u32).unwrap();
}

pub fn consume<R: tauri::Runtime>(app: &AppHandle<R>) {
    app.listen("frontend-ready", |_event| {});
    app.listen_any("sync-state", |_event| {});
    app.once("rust-internal-tick", |_event| {});
}
